gtk4/src/subclass/entry_buffer.rs

```diff
@@ -0,0 +1,286 @@
+
+        let klass = class.as_mut();
+        klass.delete_text = Some(entry_buffer_delete_text::<T>);
+        klass.deleted_text = Some(entry_buffer_deleted_text::<T>);
+        klass.get_length = Some(entry_buffer_get_length::<T>);
+        klass.get_text = Some(entry_buffer_get_text::<T>);
+        klass.insert_text = Some(entry_buffer_insert_text::<T>);
+        klass.inserted_text = Some(entry_buffer_inserted_text::<T>);
+    }
+}
+
+unsafe extern "C" fn entry_buffer_delete_text<T: EntryBufferImpl>(
+    ptr: *mut ffi::GtkEntryBuffer,
+    position: u32,
+    n_chars: u32,
+) -> u32 {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<EntryBuffer> = from_glib_borrow(ptr);
+
+    let n_chars = if n_chars == u32::MAX {
+        None
+    } else {
+        Some(n_chars)
+    };
+
+    imp.delete_text(wrap.unsafe_cast_ref(), position, n_chars)
+}
+
+unsafe extern "C" fn entry_buffer_deleted_text<T: EntryBufferImpl>(
+    ptr: *mut ffi::GtkEntryBuffer,
+    position: u32,
+    n_chars: u32,
+) {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<EntryBuffer> = from_glib_borrow(ptr);
+
+    let n_chars = if n_chars == u32::MAX {
+        None
+    } else {
+        Some(n_chars)
+    };
+
+    imp.deleted_text(wrap.unsafe_cast_ref(), position, n_chars)
+}
+
+static GET_TEXT_QUARK: Lazy<glib::Quark> =
+    Lazy::new(|| glib::Quark::from_string("gtk4-rs-subclass-entry-buffer-text"));
+
+unsafe extern "C" fn entry_buffer_get_text<T: EntryBufferImpl>(
+    ptr: *mut ffi::GtkEntryBuffer,
+    n_bytes: *mut usize,
+) -> *const libc::c_char {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<EntryBuffer> = from_glib_borrow(ptr);
+
+    let ret = imp.get_text(wrap.unsafe_cast_ref());
+    *n_bytes = ret.len();
+    // Ensures that the returned text stays alive for as long as
+    // the entry buffer instance
+    let fullptr = ret.to_glib_full();
+    wrap.set_qdata(*GET_TEXT_QUARK, fullptr);
+    fullptr
+}
+
+unsafe extern "C" fn entry_buffer_get_length<T: EntryBufferImpl>(
+    ptr: *mut ffi::GtkEntryBuffer,
+) -> u32 {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<EntryBuffer> = from_glib_borrow(ptr);
+
+    imp.get_length(wrap.unsafe_cast_ref())
+}
+
+unsafe extern "C" fn entry_buffer_insert_text<T: EntryBufferImpl>(
+    ptr: *mut ffi::GtkEntryBuffer,
+    position: u32,
+    charsptr: *const libc::c_char,
+    n_chars: u32,
+) -> u32 {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<EntryBuffer> = from_glib_borrow(ptr);
+    let text: Borrowed<GString> = from_glib_borrow(charsptr);
+
+    let chars =
+        if n_chars != u32::MAX {
+            let pos_end = text.char_indices().take(n_chars as usize).next().expect(
```

Comment:
I think this is almost correct.

```suggestion
            let mut iter = text.char_indices().skip(n_chars as usize);
            iter.next().expect("blabla"); // This would fail if there are no `n_chars` characters
            let pos_end = iter.next().unwrap_or_else(|| text.len()); // If `n_chars` is the number of characters then there is no next character and you should take the whole string
```

Reply:
Same in the other places

Reply:
Why didn't you just paste the code I wrote above? :) Now it's wrong. `pos_end` is again the byte index of the start of the last character. But you want the byte index of the start of the next character (or the length of the whole bytes slice).

Note that I called `next()` twice above and have `unwrap_or_else()` in there too.

Reply:
This code is apparently so complicated that it needs a test :) Maybe refactor it into a helper function in this file and write a few tests around it with some mixed ASCII/UTF-8 strings.

Reply:
Makes sense; I will do that

Reply:
> Why didn't you just paste the code I wrote above? :) Now it's wrong. `pos_end` is again the byte index of the start of the last character. But you want the byte index of the start of the next character (or the length of the whole bytes slice).
> 
> Note that I called `next()` twice above and have `unwrap_or_else()` in there too.

The double .next() wasn't giving the expected results, see the results of the tests I have added
```
thread 'subclass::entry_buffer::test::n_chars_max_length_ascii' panicked at 'assertion failed: `(left == right)`
  left: `"gtk-rs "`,
 right: `"gtk-rs"`', gtk4/src/subclass/entry_buffer.rs:295:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- subclass::entry_buffer::test::n_chars_max_length_utf8 stdout ----
thread 'subclass::entry_buffer::test::n_chars_max_length_utf8' panicked at 'assertion failed: `(left == right)`
  left: `"👨👩👧"`,
 right: `"👨👩"`', gtk4/src/subclass/entry_buffer.rs:310:9

---- subclass::entry_buffer::test::n_chars_max_length_utf8_ascii stdout ----
thread 'subclass::entry_buffer::test::n_chars_max_length_utf8_ascii' panicked at 'assertion failed: `(left == right)`
  left: `"👨g👩"`,
 right: `"👨g"`', gtk4/src/subclass/entry_buffer.rs:317:9

```
As you can see, there's always an extra character than the expected result

Reply:
Indeed, the skip always skips over that so one more is not needed.